grouper/fe/handlers/permission_request.py

```diff
@@ -0,0 +1,141 @@
+from grouper.user_group import get_groups_by_user
+
+
+def _build_form(request, data):
+    """Build the permission request form given the request and POST data.
+
+    Normally all fields of the form will be editable.  But if the URL
+    locks down a specific value for the group, permission, or argument,
+    then the specified fields will display those values and will be
+    grayed out and not editable.
+
+    """
+    session = request.session
+    current_user = request.current_user
+
+    def pairs(seq):
+        return [(item, item) for item in seq]
+
+    form = PermissionRequestForm(data)
+
+    group_names = {g.groupname for g, e in get_groups_by_user(session, current_user)}
+    args_by_perm = get_grantable_permissions(session, settings.restricted_ownership_permissions)
+    permission_names = {p for p in args_by_perm}
+
+    group_param = request.get_argument("group", None)
+    if group_param is not None:
+        if group_param not in group_names:
+            raise HTTPError(
+                status_code=404, reason="the group name in the URL is not one you belong to"
+            )
+        form.group_name.choices = pairs([group_param])
+        form.group_name.render_kw = {"readonly": "readonly"}
+    else:
+        form.group_name.choices = pairs([""] + sorted(group_names))
+
+    permission_param = request.get_argument("permission", None)
+    if permission_param is not None:
+        if permission_param not in permission_names:
+            raise HTTPError(
+                status_code=404, reason="an unrecognized permission is specified in the URL"
+            )
+        form.permission_name.choices = pairs([permission_param])
+        form.permission_name.render_kw = {"readonly": "readonly"}
+    else:
+        form.permission_name.choices = pairs([""] + sorted(permission_names))
+
+    argument_param = request.get_argument("argument", "")
+    if argument_param:
+        form.argument.render_kw = {"readonly": "readonly"}
+        form.argument.data = argument_param
+
+    return form, args_by_perm
+
+
+class PermissionRequest(GrouperHandler):
+    def get(self):
+        form, args_by_perm = _build_form(self, None)
+        self.render(
+            "permission-request.html",
+            args_by_perm_json=json.dumps(args_by_perm),
+            form=form,
+            uri=self.request.uri,
+        )
+
+    def post(self):
+        form, args_by_perm = _build_form(self, self.request.arguments)
+
+        if not form.validate():
+            return self.render(
+                "permission-request.html",
+                args_by_perm_json=json.dumps(args_by_perm),
+                form=form,
+                alerts=self.get_form_alerts(form.errors),
+            )
+
+        group = self.session.query(Group).filter(Group.groupname == form.group_name.data).first()
+        assert group is not None, "our prefilled permission should exist or we have problems"
```

Comment:
Let's throw an explicit error rather than asserting (here and below). A mismatch here is probably because someone is messing with the form data, so `self.badrequest()` seems appropriate here.

Reply:
Yes, I copied this from the old form, and thought it was a bit wonky, but didn't know which kind of error to use instead; I guess the original author thought "this will only happen to folks messing with the form so it doesn't matter"? I'll try `badrequest()`.